arch/arm/soc/atmel_sam0/samd/soc.c

```diff
@@ -0,0 +1,192 @@
+{
+	u32_t fuse = *(u32_t *)FUSES_OSC32K_CAL_ADDR;
+	u32_t calib = (fuse & FUSES_OSC32K_CAL_Msk) >> FUSES_OSC32K_CAL_Pos;
+
+	SYSCTRL->OSC32K.reg = SYSCTRL_OSC32K_CALIB(calib) |
+			      SYSCTRL_OSC32K_STARTUP(0x6u) |
+			      SYSCTRL_OSC32K_EN32K | SYSCTRL_OSC32K_ENABLE;
+
+	/* Wait for the oscillator to stabalise. */
+	while (!SYSCTRL->PCLKSR.bit.OSC32KRDY) {
+	}
+}
+
+static void dfll_init(void)
+{
+	/* No prescaler */
+	GCLK->GENDIV.reg = GCLK_GENDIV_ID(1) | GCLK_GENDIV_DIV(0);
+	wait_gclk_synchronization();
+
+	/* Route XOSC32K to GCLK1 */
+	GCLK->GENCTRL.reg =
+	    GCLK_GENCTRL_ID(1) | GCLK_GENCTRL_SRC_XOSC32K | GCLK_GENCTRL_GENEN;
+	wait_gclk_synchronization();
+
+	/* Route GCLK1 to multiplexer 1 */
+	GCLK->CLKCTRL.reg =
+	    GCLK_CLKCTRL_ID(0) | GCLK_CLKCTRL_GEN_GCLK1 | GCLK_CLKCTRL_CLKEN;
+	wait_gclk_synchronization();
+
+	SYSCTRL->DFLLCTRL.reg = SYSCTRL_DFLLCTRL_ENABLE;
+	while (!SYSCTRL->PCLKSR.bit.DFLLRDY) {
+	}
+
+	u32_t mul = (SOC_ATMEL_SAM0_MCK_FREQ_HZ +
+		     SOC_ATMEL_SAM0_XOSC32K_FREQ_HZ / 2) /
+		    SOC_ATMEL_SAM0_XOSC32K_FREQ_HZ;
+
+	SYSCTRL->DFLLMUL.reg = SYSCTRL_DFLLMUL_CSTEP(31) |
+			       SYSCTRL_DFLLMUL_FSTEP(511) |
+			       SYSCTRL_DFLLMUL_MUL(mul);
+	while (!SYSCTRL->PCLKSR.bit.DFLLRDY) {
+	}
+
+	SYSCTRL->DFLLCTRL.reg |= SYSCTRL_DFLLCTRL_MODE |
+				 SYSCTRL_DFLLCTRL_WAITLOCK |
+				 SYSCTRL_DFLLCTRL_QLDIS;
+	while (!SYSCTRL->PCLKSR.bit.DFLLRDY) {
+	}
+
+	/* Enable the DFLL */
+	SYSCTRL->DFLLCTRL.bit.ENABLE = 1;
+
+	while (!SYSCTRL->PCLKSR.bit.DFLLLCKC || !SYSCTRL->PCLKSR.bit.DFLLLCKF) {
+	}
+
+	while (!SYSCTRL->PCLKSR.bit.DFLLRDY) {
+	}
+}
+
+static void osc8m_init(void)
+{
+	/* Turn off the prescaler */
+	SYSCTRL->OSC8M.bit.PRESC = SYSCTRL_OSC8M_PRESC_0_Val;
+	SYSCTRL->OSC8M.bit.ONDEMAND = 0;
+}
+
+static void gclks_init(void)
+{
+	/* DFLL/1 -> GCLK0 */
+	GCLK->GENDIV.reg = GCLK_GENDIV_ID(0) | GCLK_GENDIV_DIV(0);
+	wait_gclk_synchronization();
+
+	GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(0) | GCLK_GENCTRL_SRC_DFLL48M |
+			    GCLK_GENCTRL_IDC | GCLK_GENCTRL_GENEN;
+	wait_gclk_synchronization();
+
+	/* OSC8M/1 -> GCLK3 */
+	GCLK->GENCTRL.reg =
+	    GCLK_GENCTRL_ID(3) | GCLK_GENCTRL_SRC_OSC8M | GCLK_GENCTRL_GENEN;
+	wait_gclk_synchronization();
+
+	/* OSC32K/32 -> GCLK2 */
+	GCLK->GENDIV.reg = GCLK_GENDIV_ID(2) | GCLK_GENDIV_DIV(32 - 1);
+	wait_gclk_synchronization();
+
+	GCLK->GENCTRL.reg =
+	    GCLK_GENCTRL_ID(2) | GCLK_GENCTRL_SRC_OSC32K | GCLK_GENCTRL_GENEN;
+	wait_gclk_synchronization();
+}
+
+static void dividers_init(void)
+{
+	/* Set the CPU, APBA, B, and C dividers */
+	PM->CPUSEL.reg = PM_CPUSEL_CPUDIV_DIV1;
+	PM->APBASEL.reg = PM_APBASEL_APBADIV_DIV1_Val;
+	PM->APBBSEL.reg = PM_APBBSEL_APBBDIV_DIV1_Val;
+	PM->APBCSEL.reg = PM_APBCSEL_APBCDIV_DIV1_Val;
+
+	/* TODO(mlhx): enable clock failure detection? */
+}
+
+static int atmel_samd_init(struct device *arg)
+{
+	u32_t key;
+
+	ARG_UNUSED(arg);
+
+	key = irq_lock();
+
+	_ClearFaults();
+
+	flash_waitstates_init();
+	xosc_init();
+	xosc32k_init();
```

Comment:
It looks like this code is Arduino Zero specific. It assumes that external 32k crystal oscillator is connected. This doesn't have to be the case for another board. We need to be able to configure this via Kconfig option. Please have a look at how it's done for SAM E70 (Kconfig.soc, soc.c in arch/arm/soc/atmel_sam/same70/).

Reply:
Done.  I've added support for enabling XOSC32K and a option to pick between XOSC32K, XOSC, and the internal OSC8M as clock source.  I've tested XOSC32K and OSC8M and both work fine.

Reply:
Oh, and the three boards that I spot-checked all use XOSC32K so I made it the default clock source.